src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs

```diff
@@ -515,11 +497,67 @@ protected void OnError(ErrorEventArgs e)
         [SuppressMessage("Microsoft.Security", "CA2109:ReviewVisibleEventHandlers", MessageId = "0#", Justification = "Changing from protected to private would be a breaking change")]
         protected void OnRenamed(RenamedEventArgs e)
         {
-            RenamedEventHandler renamedHandler = _onRenamedHandler;
-            if (renamedHandler != null)
+            _onRenamedHandler?.Invoke(this, e);
+        }
+
+        public WaitForChangedResult WaitForChanged(WatcherChangeTypes changeType) => 
+            WaitForChanged(changeType, Timeout.Infinite);
+
+        public WaitForChangedResult WaitForChanged(WatcherChangeTypes changeType, int timeout)
+        {
+            // The full framework implementation doesn't do any argument validation, so
+            // none is done here, either.
+
+            // Create a TCS, used as a convenient way to communicate a result from
+            // one thread to another blocked thread.
+            var tcs = new TaskCompletionSource<WaitForChangedResult>();
+            
+            // Create the event handlers to use to complete the task when an appropriate event occurs.
+            FileSystemEventHandler fseh = (s, e) =>
+            {
+                if ((e.ChangeType & changeType) != 0)
+                    tcs.TrySetResult(new WaitForChangedResult(e.ChangeType, e.Name, null, false));
+            };
+            RenamedEventHandler reh = (s, e) =>
+            {
+                if ((e.ChangeType & changeType) != 0)
+                    tcs.TrySetResult(new WaitForChangedResult(e.ChangeType, e.Name, e.OldName, false));
+            };
+
+            // Register the event handlers based on what events are desired.  The full framework
+            // doesn't register for the Error event, so this doesn't either.
+            if ((changeType & WatcherChangeTypes.Created) != 0) Created += fseh;
+            if ((changeType & WatcherChangeTypes.Deleted) != 0) Deleted += fseh;
+            if ((changeType & WatcherChangeTypes.Changed) != 0) Changed += fseh;
+            if ((changeType & WatcherChangeTypes.Renamed) != 0) Renamed += reh;
+            try
             {
-                renamedHandler(this, e);
+                // Enable the FSW if it wasn't already.
+                bool wasEnabled = EnableRaisingEvents;
+                if (!wasEnabled)
+                {
+                    EnableRaisingEvents = true;
+                }
+
+                // Block until an appropriate event arrives or until we timeout.
+                tcs.Task.Wait(timeout);
+
+                // Reset the enabled state to what it was.
+                EnableRaisingEvents = wasEnabled;
```

Comment:
It seems odd to me that it will re set EnableRaisingEvents to false if that's what it was before the method call. Is this how it worked in Desktop?